sdk/opendp/whitenoise/synthesizers/pytorch/nn/dpctgan.py

```diff
@@ -0,0 +1,331 @@
+import numpy as np
+import torch
+from torch import optim
+from torch.nn import functional
+import torch.nn as nn
+import torch.utils.data
+from torch.nn import BatchNorm1d, Dropout, LeakyReLU, Linear, Module, ReLU, Sequential,Sigmoid
+from torch.nn import functional as F
+from opendp.whitenoise.synthesizers.base import SDGYMBaseSynthesizer
+
+import ctgan
+from ctgan.transformer import DataTransformer
+from ctgan.conditional import ConditionalGenerator 
+from ctgan.models import Discriminator, Generator
+from ctgan.sampler import Sampler 
+
+from ctgan import CTGANSynthesizer
+
+import torchdp
+from torchdp import autograd_grad_sample
+from torchdp import PrivacyEngine, utils
+
+
+
+def __init__(self, input_dim, dis_dims, loss, pack):
+        super(Discriminator, self).__init__()
+        torch.cuda.manual_seed(0)
+        torch.manual_seed(0)
+        
+        dim = input_dim * pack
+      #  print ('now dim is {}'.format(dim))
+        self.pack = pack
+        self.packdim = dim
+        seq = []
+        for item in list(dis_dims):
+            seq += [Linear(dim, item), LeakyReLU(0.2), Dropout(0.5)]
+            dim = item
+
+        seq += [Linear(dim, 1)]
+        if loss == 'cross_entropy':
+            print ("this should happen")
+            seq += [Sigmoid()]
+        self.seq = Sequential(*seq)
+        
+Discriminator.__init__ = __init__
+
+
+
+
+
+
+
+class DPCTGAN(SDGYMBaseSynthesizer, CTGANSynthesizer):
```

Comment:
We should not need this class anymore. Or at least, it should not have the Synthesizers. The GAN should be constructed from here and passed to PytorchDPSynthesizer

Reply:
oh, right, SDGYMBaseSynthesizer isn't needed here.  I think CTGANSynthesizer is still needed, because DPCTGAN is extending the CTGANSynthesizer import from ctgan.